Feature/ci unique openwrt builder names

rmelotte opened the pull request:
- Make it possible to choose the image tag. While doing that, switch to using command line options instead of files when calling the build script.
- Use unique names for the openwrt builder, so that we can safely run multiple jobs at the same time. 

This should make the CI pipelines to be picked up faster, since it will no longer be limited to 1 job at a time (once we make the related change to the runner).

rmelotte committed:
tools: openwrt: build.sh: switch to options instead of variables

Using command line options makes it easier to control what the build
script does.

Add command line options to build.sh, and drop the "args" files.

The CI configuration is adapted to use the new options:

- drop the target file, add the '-d' option.
- while we're at it, use '-t' to give a unique tags to the containers,
  so that the job can be run several times at the same time on the
  same runner.
- also add an additional pattern to the grep command to display
  Docker's behavior (using cache or building).

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: docker: make the openwrt build script stop at errors

When using the build script in CI, it's essential that it either:
- records errors and exits with a corresponding exit code at the end,
or
- fail at the first error.

Otherwise, the CI could report a successful job while in reality it
failed.

Add the '-e' option to the build script, so that it stops with an
error code at the first failing command.

Also add the '-v' option to the last cp command, to be verbose about
what is being copied.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

rmelotte committed:
tools: docker: openwrt: move the source directory to home

The source used to compile prplMesh was previously mounted as a volume
under '/'. This was possible when running the container manually. When
using in CI, we cannot always control where the source will be
cloned (it depends on the runner configuration to allow that or not).

Moreover, the openwrt user inside the container cannot add a symlink
or copy files to '/', because it doesn't have the permission to write
there.

scripts/build.sh: look for the source in prplMesh_source under the
user home instead of '/'.

tools/docker/builder/openwrt/build.sh: mount the source volume in
prplMesh_source in the user home.

While we're at it, remove the TARGET_PLATFORM_TYPE patching from both
files, because it's not needed anymore.

With these changes, the CI does not have to rely on where the source
will be cloned, and can just add a symlink in
/home/openwrt/prplMesh_source pointing to the source (for example).

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

tomereli commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -3,75 +3,123 @@
+. "${rootdir}/tools/docker/functions.sh"
+
+usage() {
+    echo "usage: $(basename $0) [-ipt]"
```
short options are missing some of the long options

rmelotte replied:
Done.

tomereli commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -3,75 +3,123 @@
+        dbg "image tag not set, using default value $image_tag"
+    fi
+
+    export OPENWRT_REPOSITORY
```
Why do we need empty exports?

rmelotte replied:
They need to be exported for `docker build` to be able to use them. The other option is to use `docker build --build-arg OPTION=$OPTION` everywhere (i.e. specifying the arg value explicitly), but I didn't want to type the args twice everywhere.
If there is a good reason to do it I can still change it.

tomereli commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -3,75 +3,123 @@
+    esac
+
+    if [ -z "$OPENWRT_REPOSITORY" ] ; then
+        OPENWRT_REPOSITORY=https://git.prpl.dev/prplmesh/prplwrt.git
```
I prefer to declare all defaults outside of `main()` where we put `VERBOSE`, then simply update them directly in the `getopt` loop if selected.

rmelotte replied:
Done except for "image_tag", which depend on other options.

tomereli commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -3,75 +3,123 @@
+        esac
+    done
+
+    case "$TARGET_DEVICE" in
```
I prefer to add this `case` to the `while` loop above and treat it like any other argument - by overriding defaults which are declared outside of `main`.

rmelotte replied:
Done.

tomereli commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -3,75 +3,123 @@
+           "$scriptdir/"
+}
+
+build_prplmesh() {
```
I think it is better to explicitly pass parameters to this function and use $1 $2 etc. into local parameters (nitpick)

tomereli commented on .gitlab-ci.yml:
```diff
@@ -49,7 +49,7 @@ run-tests:
   script:
     - cd tools/docker/builder/openwrt
-    - ./build.sh targets/$TARGET.args 2>&1 | tee build.log | grep -E '^make\[[12]\]|^Step' --color=never --line-buffered
+    - ./build.sh -v -d "$TARGET" -t "prplmesh-builder-$TARGET:$CI_COMMIT_SHORT_SHA-$CI_PIPELINE_ID" 2>&1 | tee build.log | grep '^make\[[12]\]\|^Step\|^ --->' --color=never --line-buffered
```
I think its better to change `$TARGET` to `$TARGET_DEVICE` since it will be more readable this way IMO.

rmelotte replied:
Done in a separate commit, since the variables are already in master, and the changes in this PR are not related.

tomereli reviewed (approved):
Mainly nitpicks so already approving, please fix the comments or reply before you merge.

rmelotte committed:
.gitlab-ci.yml: rename TARGET and TARGET_DEVICE

The build script uses "TARGET" to name the target system (e.g.: mvebu)
while the gitlab-ci.yml file uses TARGET to name a target
profile (e.g.: turris-omnia).

.gitlab-ci.yml: rename TARGET_DEVICE to TARGET_DEVICE_NAME, and rename
TARGET to TARGET_DEVICE, to align with the build script.

Signed-off-by: Raphaël Mélotte <raphael.melotte@mind.be>

tomereli commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -56,7 +57,19 @@ main() {
+			;;
+		esac
+		;;
+	    -f | --prpl-feed)          PRPL_FEED="$2"; shift; shift ;;
```
missing a space..

rmelotte replied:
Crap, tabs vs spaces again. Let me untabify it.

tomereli commented on tools/docker/builder/openwrt/build.sh:
```diff
@@ -118,5 +113,8 @@ main() {
 VERBOSE=false
+OPENWRT_REPOSITORY='https://git.prpl.dev/prplmesh/prplwrt.git'
+OPENWRT_VERSION='9d2efd'
+PRPL_FEED='https://git.prpl.dev/prplmesh/iwlwav.git^06a0126d5fb53b1d65bad90757a5f9f5f77419ca'
```
Great!

tomereli reviewed (approved):
LGTM

rmelotte commented:
Don't merge yet, the last commit is missing something (the fixup also needs to be squashed ofc).

rmelotte commented:
Rebased on master, squashed the fixups.
There are 2 open review comments, other than I think it can be merged.
